### ISSUE TITLE
Native Store SailSourceModel concurrency bug

### ISSUE THREAD

**Issue by jeenbroekstra** [eclipse/rdf4j#693]
This is a followup of issue #680 . The Native Store internally uses a MemoryOverflowModel to cope with large transaction sizes by caching data to disk. This in turn uses a SailSourceModel to physically store and retrieve the cached data. SailSourceModel does not properly synchronize its public access methods, which means that when concurrent txns try to access it, it can lead to intermittent exceptions on manipulation or retrieval of data. Example stacktrace:

```
Exception in thread "Thread-2" java.lang.NullPointerException
	at org.eclipse.rdf4j.sail.nativerdf.SailSourceModel.dataset(SailSourceModel.java:397)
	at org.eclipse.rdf4j.sail.nativerdf.SailSourceModel.contains(SailSourceModel.java:227)
	at org.eclipse.rdf4j.sail.nativerdf.MemoryOverflowModel.contains(MemoryOverflowModel.java:111)
	at org.eclipse.rdf4j.model.impl.FilteredModel.contains(FilteredModel.java:147)
	at org.eclipse.rdf4j.model.impl.AbstractModel.isEmpty(AbstractModel.java:45)
	at org.eclipse.rdf4j.sail.base.SailDatasetImpl.union(SailDatasetImpl.java:336)
	at org.eclipse.rdf4j.sail.base.SailDatasetImpl.getStatements(SailDatasetImpl.java:305)
	at org.eclipse.rdf4j.sail.base.DelegatingSailDataset.getStatements(DelegatingSailDataset.java:74)
	at org.eclipse.rdf4j.sail.base.DelegatingSailDataset.getStatements(DelegatingSailDataset.java:74)
	at org.eclipse.rdf4j.sail.base.SailSourceConnection.remove(SailSourceConnection.java:720)
	at org.eclipse.rdf4j.sail.base.SailSourceConnection.removeStatement(SailSourceConnection.java:551)
	at org.eclipse.rdf4j.sail.helpers.AbstractSailConnection.removeStatements(AbstractSailConnection.java:491)
	at org.eclipse.rdf4j.repository.sail.SailRepositoryConnection.removeWithoutCommit(SailRepositoryConnection.java:382)
	at org.eclipse.rdf4j.repository.base.AbstractRepositoryConnection.remove(AbstractRepositoryConnection.java:587)
```

**Comment by jeenbroekstra** [comment]
Possible alternative to synchronizing all public accessors is to synchronize only a few specific internal methods. This may improve performance.

**Comment by ansell** [comment]
Multi Read/Single Write locks may have higher throughput than full synchronisation

**Comment by jeenbroekstra** [comment]
Good point. I'm sticking with synchronized for now, just to make sure that this at least fixes the bug. If I have time I'll look at a more sophisticated lock to improve performance.

**Comment by jeenbroekstra** [comment]
As a further thought: the reason these kinds of issues do not show up in our current test rig is that MemoryOverflowModel does not start syncing to disk until a txn hits the 10k statements boundary. I'll look at making that number for configurable for testing and/or including at least one test that works with sufficiently large data to trigger disk caching.

**Comment by jeenbroekstra** [comment]
I ran a few tests with a draft implementation using a multi-read/single-write lock manager, but I couldn't get it stable (exceptions as well as inconsistent results), and moreover its performance was similar (or even worse) to the simple 'synchronize-all' approach.

**Comment by jeenbroekstra** [comment]
@rknowles9 I have attached a 
[new snapshot build of rdf4j-server](https://github.com/eclipse/rdf4j/files/648102/rdf4j-server.war.zip) with a fix for this issue in place. We're going to release a patch release tomorrow, but if you have time before then to give this a spin and see if it solves your problem, then that'd be great.

**Comment by rknowles9** [comment]
I grabbed the snapshot and tested. I'm getting the same irregular results- sometimes 2 contexts are loaded, sometimes 1, and sometimes 0, I haven't set up a massive repeat of tests to be sure of this, but it feels like I get more correct results when I test from Eclipse than when I test from command line, and it feels like I get more error results when I keep checking during the run. By this, I mean I sometimes watch for results by looking at the Workbench and just refreshing the screen again and again while waiting for the test to complete. My sense is more failures doing this than when I just run the test, wait for it to end, and then check on Workbench. But in all of the cases (eclipse-check, eclipse-don't check, command-check, command-don't check) I get the irregular results.

I'm trying again to attach my updated test. I did make another change in pom.xml to use rdf4j 2.2-SNAPSHOT after packaging this.

[Issue680test,v2.zip](https://github.com/eclipse/rdf4j/files/649206/Issue680test.v2.zip)

**Comment by jeenbroekstra** [comment]
The problem appears to be specific to the native store, but the odd thing is that I can only reproduce it when accessing a native store remotely (that is, over http). My previous tests were using a local native store, and the fixes I did in SailSourceModel appeared to fix the issue in that scenario. 

I have also determined that the clearing of contexts (or the clearing of summary data) that your test code does has no influence - I can reproduce the issue with simplified code that just does two concurrent uploads only.

**Comment by jeenbroekstra** [comment]
Further data point: I have found that for two concurrent transactions the native store creates two separate SailSourceModels for the overflow, as expected. When inspecting the temporary dirs afterwards, I found that in the cases where the upload had gone wrong (only showing one context), one of these temp dirs still had a `txn-status` file set to `ACTIVE`. It appears that intermittently the model is not correctly synced to the main store and closed.
